File: ios/Runner/Crypto/CryptoEngine.swift
import Foundation
import Security

/// Cooperative cancellation flag checked between chunks of a streaming operation.
final class CancellationToken {
    private var _cancelled = false
    private let lock = NSLock()
    func cancel() { lock.lock(); _cancelled = true; lock.unlock() }
    var isCancelled: Bool { lock.lock(); defer { lock.unlock() }; return _cancelled }
}

typealias ProgressCallback = (_ bytesProcessed: Int64, _ totalBytes: Int64) -> Void

/// Hybrid RSA-OAEP(SHA-256) + AES-256-GCM engine. Mirrors the Kotlin
/// `EncryptionEngine` exactly so the two platforms produce byte-identical
/// containers; see `CryptoFormat` for the shared layout.
///
/// All streaming methods use `InputStream`/`OutputStream` with a fixed-size
/// buffer (`CryptoFormat.streamBufferSize`), never loading a whole file into
/// memory, so RAM usage stays constant regardless of file size.
final class CryptoEngine {

    private let keyManager = KeychainKeyManager()

    // MARK: Key management

    func generateRSAKeyPair(alias: String) throws -> [UInt8] {
        try keyManager.generateKeyPair(alias: alias)
    }

    func importPublicKey(alias: String, spkiDer: [UInt8]) throws {
        try keyManager.importPublicKey(alias: alias, spkiDer: spkiDer)
    }

    func importPrivateKey(alias: String, pkcs8Der: [UInt8]) throws {
        try keyManager.importPrivateKey(alias: alias, pkcs8Der: pkcs8Der)
    }

    func exportPublicKey(alias: String) throws -> [UInt8] {
        try keyManager.exportPublicKeyDer(alias: alias)
    }

    func deleteKey(alias: String) {
        keyManager.deleteKeyPair(alias: alias)
    }

    // MARK: File streaming

    func encryptFile(inputPath: String, outputPath: String, publicKeyAlias: String,
                      token: CancellationToken, onProgress: ProgressCallback?) throws {
        let fm = FileManager.default
        guard fm.fileExists(atPath: inputPath) else {
            throw CryptoError.ioFailure("Input file does not exist: \(inputPath)")
        }
        let totalBytes = (try? fm.attributesOfItem(atPath: inputPath)[.size] as? Int64) ?? 0

        let publicKey = try keyManager.getPublicKey(alias: publicKeyAlias)
        let aesKey = randomBytes(CryptoFormat.aesKeyLength)
        let iv = randomBytes(CryptoFormat.gcmIvLength)
        let encryptedAesKey = try rsaEncryptKey(aesKey, with: publicKey)

        guard let input = InputStream(fileAtPath: inputPath) else {
            throw CryptoError.ioFailure("Unable to open input file for reading")
        }
        let tmpPath = outputPath + ".tmp"
        fm.createFile(atPath: tmpPath, contents: nil)
        guard let output = OutputStream(toFileAtPath: tmpPath, append: false) else {
            throw CryptoError.ioFailure("Unable to open output file for writing")
        }
        input.open(); output.open()
        defer { input.close(); output.close() }

        do {
            try writeHeader(output, encryptedAesKey: encryptedAesKey, iv: iv)
            let cipher = try StreamingAESGCM(operation: CCOperation(kCCEncrypt), key: aesKey, iv: iv)
            try pump(input: input, output: output, totalBytes: totalBytes ?? 0,
                     token: token, onProgress: onProgress) { chunk in
                try cipher.update(chunk, isEncrypt: true)
            }
            let tag = try cipher.finalizeEncrypt()
            try write(output, tag)
            try moveIntoPlace(tmpPath: tmpPath, finalPath: outputPath)
        } catch {
            try? fm.removeItem(atPath: tmpPath)
            throw mapError(error, defaultCase: .encryptionFailed)
        }
    }

    func decryptFile(inputPath: String, outputPath: String, privateKeyAlias: String,
                      token: CancellationToken, onProgress: ProgressCallback?) throws {
        let fm = FileManager.default
        guard fm.fileExists(atPath: inputPath) else {
            throw CryptoError.ioFailure("Input file does not exist: \(inputPath)")
        }
        let totalBytes = (try? fm.attributesOfItem(atPath: inputPath)[.size] as? Int64) ?? 0
        let privateKey = try keyManager.getPrivateKey(alias: privateKeyAlias)

        guard let input = InputStream(fileAtPath: inputPath) else {
            throw CryptoError.ioFailure("Unable to open input file for reading")
        }
        input.open()
        defer { input.close() }

        let (encryptedAesKey, iv) = try readHeader(input)
        let aesKey = try rsaDecryptKey(encryptedAesKey, with: privateKey)
        let headerSize = CryptoFormat.magicBytes.count + 4 + encryptedAesKey.count + CryptoFormat.gcmIvLength
        let ciphertextAndTagSize = (totalBytes ?? 0) - Int64(headerSize)
        let ciphertextSize = ciphertextAndTagSize - Int64(CryptoFormat.gcmTagLength)
        guard ciphertextSize >= 0 else {
            throw CryptoError.invalidFormat("File is smaller than the declared header")
        }

        let tmpPath = outputPath + ".tmp"
        fm.createFile(atPath: tmpPath, contents: nil)
        guard let output = OutputStream(toFileAtPath: tmpPath, append: false) else {
            throw CryptoError.ioFailure("Unable to open output file for writing")
        }
        output.open()
        defer { output.close() }

        do {
            let cipher = try StreamingAESGCM(operation: CCOperation(kCCDecrypt), key: aesKey, iv: iv)
            try pumpFixedLength(input: input, output: output, byteCount: ciphertextSize,
                                 totalBytes: totalBytes ?? 0, token: token, onProgress: onProgress) { chunk in
                try cipher.update(chunk, isEncrypt: false)
            }
            let tag = try readExactly(input, count: CryptoFormat.gcmTagLength)
            try cipher.finalizeDecrypt(expectedTag: tag)
            try moveIntoPlace(tmpPath: tmpPath, finalPath: outputPath)
        } catch {
            try? fm.removeItem(atPath: tmpPath)
            throw mapError(error, defaultCase: .decryptionFailed)
        }
    }

    // MARK: Bytes API (delegates to the same primitives as the file API)

    func encryptBytes(_ data: [UInt8], publicKeyAlias: String) throws -> [UInt8] {
        let publicKey = try keyManager.getPublicKey(alias: publicKeyAlias)
        let aesKey = randomBytes(CryptoFormat.aesKeyLength)
        let iv = randomBytes(CryptoFormat.gcmIvLength)
        let encryptedAesKey = try rsaEncryptKey(aesKey, with: publicKey)

        var out: [UInt8] = []
        out.append(contentsOf: CryptoFormat.magicBytes)
        out.append(contentsOf: uint32BE(encryptedAesKey.count))
        out.append(contentsOf: encryptedAesKey)
        out.append(contentsOf: iv)

        do {
            let cipher = try StreamingAESGCM(operation: CCOperation(kCCEncrypt), key: aesKey, iv: iv)
            var offset = 0
            while offset < data.count {
                let end = min(offset + CryptoFormat.streamBufferSize, data.count)
                out.append(contentsOf: try cipher.update(Array(data[offset..<end]), isEncrypt: true))
                offset = end
            }
            out.append(contentsOf: try cipher.finalizeEncrypt())
            return out
        } catch {
            throw mapError(error, defaultCase: .encryptionFailed)
        }
    }

    func decryptBytes(_ data: [UInt8], privateKeyAlias: String) throws -> [UInt8] {
        let privateKey = try keyManager.getPrivateKey(alias: privateKeyAlias)
        var offset = 0
        guard data.count >= CryptoFormat.magicBytes.count + 4 else {
            throw CryptoError.invalidFormat("Data too short to contain a header")
        }
        guard Array(data[0..<CryptoFormat.magicBytes.count]) == CryptoFormat.magicBytes else {
            throw CryptoError.invalidFormat("Unrecognized header; expected '\(CryptoFormat.magic)'")
        }
        offset = CryptoFormat.magicBytes.count
        let keyLen = Int(beUInt32(Array(data[offset..<offset + 4]))); offset += 4
        guard keyLen > 0, keyLen <= 4096, offset + keyLen + CryptoFormat.gcmIvLength <= data.count else {
            throw CryptoError.invalidFormat("Invalid encrypted key length in header")
        }
        let encryptedAesKey = Array(data[offset..<offset + keyLen]); offset += keyLen
        let iv = Array(data[offset..<offset + CryptoFormat.gcmIvLength]); offset += CryptoFormat.gcmIvLength

        let aesKey = try rsaDecryptKey(encryptedAesKey, with: privateKey)
        guard data.count - offset >= CryptoFormat.gcmTagLength else {
            throw CryptoError.invalidFormat("Data too short to contain a GCM tag")
        }
        let ciphertextEnd = data.count - CryptoFormat.gcmTagLength
        do {
            let cipher = try StreamingAESGCM(operation: CCOperation(kCCDecrypt), key: aesKey, iv: iv)
            var plain: [UInt8] = []
            var cur = offset
            while cur < ciphertextEnd {
                let end = min(cur + CryptoFormat.streamBufferSize, ciphertextEnd)
                plain.append(contentsOf: try cipher.update(Array(data[cur..<end]), isEncrypt: false))
                cur = end
            }
            let tag = Array(data[ciphertextEnd..<data.count])
            try cipher.finalizeDecrypt(expectedTag: tag)
            return plain
        } catch {
            throw mapError(error, defaultCase: .decryptionFailed)
        }
    }

    // MARK: RSA-OAEP-SHA256 (Security framework)

    private func rsaEncryptKey(_ aesKey: [UInt8], with publicKey: SecKey) throws -> [UInt8] {
        var error: Unmanaged<CFError>?
        guard let result = SecKeyCreateEncryptedData(
            publicKey, .rsaEncryptionOAEPSHA256, Data(aesKey) as CFData, &error
        ) else {
            throw CryptoError.encryptionFailed((error?.takeRetainedValue() as Error?)?.localizedDescription ?? "RSA-OAEP encryption failed")
        }
        return [UInt8](result as Data)
    }

    private func rsaDecryptKey(_ encrypted: [UInt8], with privateKey: SecKey) throws -> [UInt8] {
        var error: Unmanaged<CFError>?
        guard let result = SecKeyCreateDecryptedData(
            privateKey, .rsaEncryptionOAEPSHA256, Data(encrypted) as CFData, &error
        ) else {
            throw CryptoError.decryptionFailed((error?.takeRetainedValue() as Error?)?.localizedDescription ?? "RSA-OAEP decryption failed")
        }
        return [UInt8](result as Data)
    }

    // MARK: Streaming helpers

    private func writeHeader(_ output: OutputStream, encryptedAesKey: [UInt8], iv: [UInt8]) throws {
        try write(output, CryptoFormat.magicBytes)
        try write(output, uint32BE(encryptedAesKey.count))
        try write(output, encryptedAesKey)
        try write(output, iv)
    }

    private func readHeader(_ input: InputStream) throws -> ([UInt8], [UInt8]) {
        let magic = try readExactly(input, count: CryptoFormat.magicBytes.count)
        guard magic == CryptoFormat.magicBytes else {
            throw CryptoError.invalidFormat("Unrecognized file header; expected '\(CryptoFormat.magic)'")
        }
        let lenBytes = try readExactly(input, count: 4)
        let keyLen = Int(beUInt32(lenBytes))
        guard keyLen > 0, keyLen <= 4096 else {
            throw CryptoError.invalidFormat("Invalid encrypted key length in header: \(keyLen)")
        }
        let encryptedKey = try readExactly(input, count: keyLen)
        let iv = try readExactly(input, count: CryptoFormat.gcmIvLength)
        return (encryptedKey, iv)
    }

    /// Streams the remainder of `input` through `transform`, writing results to `output`.
    private func pump(input: InputStream, output: OutputStream, totalBytes: Int64,
                       token: CancellationToken, onProgress: ProgressCallback?,
                       transform: ([UInt8]) throws -> [UInt8]) throws {
        var buffer = [UInt8](repeating: 0, count: CryptoFormat.streamBufferSize)
        var processed: Int64 = 0
        while input.hasBytesAvailable {
            if token.isCancelled { throw CryptoError.cancelled }
            let read = buffer.withUnsafeMutableBytes { ptr in
                input.read(ptr.bindMemory(to: UInt8.self).baseAddress!, maxLength: CryptoFormat.streamBufferSize)
            }
            if read < 0 { throw CryptoError.ioFailure(input.streamError?.localizedDescription ?? "Read error") }
            if read == 0 { break }
            let chunk = Array(buffer[0..<read])
            try write(output, try transform(chunk))
            processed += Int64(read)
            onProgress?(processed, totalBytes)
        }
        onProgress?(max(processed, totalBytes), totalBytes)
    }

    /// Like `pump`, but stops after exactly `byteCount` input bytes (used for
    /// decryption, where the GCM tag trails the ciphertext in the same stream).
    private func pumpFixedLength(input: InputStream, output: OutputStream, byteCount: Int64, totalBytes: Int64,
                                  token: CancellationToken, onProgress: ProgressCallback?,
                                  transform: ([UInt8]) throws -> [UInt8]) throws {
        var buffer = [UInt8](repeating: 0, count: CryptoFormat.streamBufferSize)
        var remaining = byteCount
        var processed: Int64 = 0
        while remaining > 0 {
            if token.isCancelled { throw CryptoError.cancelled }
            let toRead = Int(min(Int64(CryptoFormat.streamBufferSize), remaining))
            let read = buffer.withUnsafeMutableBytes { ptr in
                input.read(ptr.bindMemory(to: UInt8.self).baseAddress!, maxLength: toRead)
            }
            if read <= 0 { throw CryptoError.invalidFormat("Unexpected end of stream while reading ciphertext") }
            let chunk = Array(buffer[0..<read])
            try write(output, try transform(chunk))
            remaining -= Int64(read)
            processed += Int64(read)
            onProgress?(processed, totalBytes)
        }
    }

    private func readExactly(_ input: InputStream, count: Int) throws -> [UInt8] {
        var result = [UInt8](repeating: 0, count: count)
        var offset = 0
        while offset < count {
            let read = result.withUnsafeMutableBytes { ptr -> Int in
                input.read(ptr.baseAddress!.advanced(by: offset), maxLength: count - offset)
            }
            if read <= 0 { throw CryptoError.invalidFormat("Unexpected end of stream while reading header") }
            offset += read
        }
        return result
    }

    private func write(_ output: OutputStream, _ bytes: [UInt8]) throws {
        guard !bytes.isEmpty else { return }
        var offset = 0
        while offset < bytes.count {
            let written = bytes.withUnsafeBytes { ptr -> Int in
                output.write(ptr.bindMemory(to: UInt8.self).baseAddress!.advanced(by: offset), maxLength: bytes.count - offset)
            }
            if written <= 0 { throw CryptoError.ioFailure(output.streamError?.localizedDescription ?? "Write error") }
            offset += written
        }
    }

    private func moveIntoPlace(tmpPath: String, finalPath: String) throws {
        let fm = FileManager.default
        if fm.fileExists(atPath: finalPath) { try? fm.removeItem(atPath: finalPath) }
        try fm.moveItem(atPath: tmpPath, toPath: finalPath)
    }

    private func randomBytes(_ count: Int) -> [UInt8] {
        var bytes = [UInt8](repeating: 0, count: count)
        _ = SecRandomCopyBytes(kSecRandomDefault, count, &bytes)
        return bytes
    }

    private func uint32BE(_ value: Int) -> [UInt8] {
        let v = UInt32(value)
        return [UInt8(v >> 24 & 0xFF), UInt8(v >> 16 & 0xFF), UInt8(v >> 8 & 0xFF), UInt8(v & 0xFF)]
    }

    private func beUInt32(_ bytes: [UInt8]) -> UInt32 {
        (UInt32(bytes[0]) << 24) | (UInt32(bytes[1]) << 16) | (UInt32(bytes[2]) << 8) | UInt32(bytes[3])
    }

    private func mapError(_ error: Error, defaultCase: (String) -> CryptoError) -> CryptoError {
        if let e = error as? CryptoError { return e }
        return defaultCase(error.localizedDescription)
    }
}

File: android/app/src/main/kotlin/com/bander/sero/crypto/CryptoEngine.kt
package com.bander.sero.crypto

import android.content.Context
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.EOFException
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStream
import java.io.OutputStream
import java.security.SecureRandom
import java.security.PrivateKey
import java.security.PublicKey
import java.util.concurrent.atomic.AtomicBoolean
import javax.crypto.AEADBadTagException
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Cancellation token passed down into long-running streaming operations.
 * `EncryptionEngine` checks [isCancelled] between chunks so a job can be
 * aborted promptly without leaving partial output mid-write when possible.
 */
class CancellationToken {
    private val cancelled = AtomicBoolean(false)
    fun cancel() = cancelled.set(true)
    val isCancelled: Boolean get() = cancelled.get()
}

/** Reports streaming progress. Values are bytes processed / total bytes (total may be -1 if unknown). */
typealias ProgressCallback = (bytesProcessed: Long, totalBytes: Long) -> Unit

/**
 * Hybrid RSA-OAEP(SHA-256) + AES-256-GCM crypto engine.
 *
 * A fresh random AES-256 key is generated per encryption operation, used to
 * encrypt the payload with AES-GCM, and is itself protected with RSA-OAEP
 * using the recipient's public key. The container format is documented in
 * [CryptoFormat] and is binary-compatible with the companion Python
 * implementation.
 *
 * All streaming methods operate with a constant-size buffer
 * ([CryptoFormat.STREAM_BUFFER_SIZE]) regardless of input size, so memory
 * usage does not grow with file size (tested conceptually up to 20GB+;
 * throughput is bound by disk/IO, not RAM).
 */
class EncryptionEngine(context: Context) {

    private val keyStore = SecureKeyStore(context)
    private val secureRandom = SecureRandom()

    // ------------------------------------------------------------------
    // Key management (delegates to SecureKeyStore)
    // ------------------------------------------------------------------

    fun generateRSAKeyPair(alias: String): ByteArray = keyStore.generateKeyPair(alias).encoded

    fun importPublicKey(alias: String, spkiDer: ByteArray): Unit {
        keyStore.importPublicKey(alias, spkiDer)
    }

    fun importPrivateKey(alias: String, pkcs8Der: ByteArray): Unit {
        keyStore.importPrivateKey(alias, pkcs8Der)
    }

    fun exportPublicKey(alias: String): ByteArray = keyStore.exportPublicKeyDer(alias)

    fun deleteKey(alias: String) {
        keyStore.deleteKeystoreKey(alias)
        keyStore.deleteImportedKeys(alias)
    }

    // ------------------------------------------------------------------
    // File streaming API
    // ------------------------------------------------------------------

    fun encryptFile(
        inputPath: String,
        outputPath: String,
        publicKeyAlias: String,
        token: CancellationToken,
        onProgress: ProgressCallback?
    ) {
        val input = File(inputPath)
        if (!input.exists() || !input.isFile) {
            throw CryptoException.IoFailure("Input file does not exist: $inputPath")
        }
        val publicKey = keyStore.resolvePublicKey(publicKeyAlias)
        val totalBytes = input.length()

        val aesKey = generateAesKey()
        val encryptedAesKey = rsaEncryptKey(aesKey, publicKey)
        val iv = ByteArray(CryptoFormat.GCM_IV_LENGTH).also { secureRandom.nextBytes(it) }

        val cipher = Cipher.getInstance(CryptoFormat.AES_TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, aesKey, GCMParameterSpec(CryptoFormat.GCM_TAG_LENGTH_BITS, iv))

        var tmpOut: File? = null
        try {
            tmpOut = File(outputPath + ".tmp")
            BufferedInputStream(FileInputStream(input), CryptoFormat.STREAM_BUFFER_SIZE).use { rawIn ->
                BufferedOutputStream(FileOutputStream(tmpOut), CryptoFormat.STREAM_BUFFER_SIZE).use { rawOut ->
                    writeHeader(rawOut, encryptedAesKey, iv)
                    streamThroughCipher(rawIn, rawOut, cipher, totalBytes, token, onProgress)
                }
            }
            if (!tmpOut.renameTo(File(outputPath))) {
                tmpOut.copyTo(File(outputPath), overwrite = true)
                tmpOut.delete()
            }
        } catch (c: CryptoException.Cancelled) {
            tmpOut?.delete()
            throw c
        } catch (t: Throwable) {
            tmpOut?.delete()
            throw CryptoException.EncryptionFailed("File encryption failed", t)
        } finally {
            aesKey.destroyQuietly()
        }
    }

    fun decryptFile(
        inputPath: String,
        outputPath: String,
        privateKeyAlias: String,
        token: CancellationToken,
        onProgress: ProgressCallback?
    ) {
        val input = File(inputPath)
        if (!input.exists() || !input.isFile) {
            throw CryptoException.IoFailure("Input file does not exist: $inputPath")
        }
        val privateKey = keyStore.resolvePrivateKey(privateKeyAlias)
        val totalBytes = input.length()

        var tmpOut: File? = null
        try {
            BufferedInputStream(FileInputStream(input), CryptoFormat.STREAM_BUFFER_SIZE).use { rawIn ->
                val (encryptedAesKey, iv) = readHeader(rawIn)
                val aesKey = rsaDecryptKey(encryptedAesKey, privateKey)
                try {
                    val cipher = Cipher.getInstance(CryptoFormat.AES_TRANSFORMATION)
                    cipher.init(Cipher.DECRYPT_MODE, aesKey, GCMParameterSpec(CryptoFormat.GCM_TAG_LENGTH_BITS, iv))

                    tmpOut = File(outputPath + ".tmp")
                    val headerSize = CryptoFormat.MAGIC_BYTES.size + 2 + encryptedAesKey.size + CryptoFormat.GCM_IV_LENGTH
                    val ciphertextBytes = totalBytes - headerSize
                    BufferedOutputStream(FileOutputStream(tmpOut), CryptoFormat.STREAM_BUFFER_SIZE).use { rawOut ->
                        streamThroughCipher(rawIn, rawOut, cipher, ciphertextBytes, token, onProgress)
                    }
                    if (!tmpOut!!.renameTo(File(outputPath))) {
                        tmpOut!!.copyTo(File(outputPath), overwrite = true)
                        tmpOut!!.delete()
                    }
                } finally {
                    aesKey.destroyQuietly()
                }
            }
        } catch (c: CryptoException.Cancelled) {
            tmpOut?.delete()
            throw c
        } catch (e: AEADBadTagException) {
            tmpOut?.delete()
            throw CryptoException.AuthenticationFailed()
        } catch (e: CryptoException) {
            tmpOut?.delete()
            throw e
        } catch (t: Throwable) {
            tmpOut?.delete()
            throw CryptoException.DecryptionFailed("File decryption failed", t)
        }
    }

    // ------------------------------------------------------------------
    // In-memory bytes API (delegates to the same streaming core so byte
    // arrays and files are always processed identically)
    // ------------------------------------------------------------------

    fun encryptBytes(data: ByteArray, publicKeyAlias: String): ByteArray {
        val publicKey = keyStore.resolvePublicKey(publicKeyAlias)
        val aesKey = generateAesKey()
        val encryptedAesKey = rsaEncryptKey(aesKey, publicKey)
        val iv = ByteArray(CryptoFormat.GCM_IV_LENGTH).also { secureRandom.nextBytes(it) }
        val cipher = Cipher.getInstance(CryptoFormat.AES_TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, aesKey, GCMParameterSpec(CryptoFormat.GCM_TAG_LENGTH_BITS, iv))
        try {
            val out = ByteArrayOutputStream(data.size + 64)
            writeHeader(out, encryptedAesKey, iv)
            streamThroughCipher(ByteArrayInputStream(data), out, cipher, data.size.toLong(), CancellationToken(), null)
            return out.toByteArray()
        } catch (t: Throwable) {
            throw CryptoException.EncryptionFailed("Byte encryption failed", t)
        } finally {
            aesKey.destroyQuietly()
        }
    }

    fun decryptBytes(data: ByteArray, privateKeyAlias: String): ByteArray {
        val privateKey = keyStore.resolvePrivateKey(privateKeyAlias)
        try {
            val input = ByteArrayInputStream(data)
            val (encryptedAesKey, iv) = readHeader(input)
            val aesKey = rsaDecryptKey(encryptedAesKey, privateKey)
            try {
                val cipher = Cipher.getInstance(CryptoFormat.AES_TRANSFORMATION)
                cipher.init(Cipher.DECRYPT_MODE, aesKey, GCMParameterSpec(CryptoFormat.GCM_TAG_LENGTH_BITS, iv))
                val out = ByteArrayOutputStream(maxOf(data.size - 64, 0))
                streamThroughCipher(input, out, cipher, input.available().toLong(), CancellationToken(), null)
                return out.toByteArray()
            } finally {
                aesKey.destroyQuietly()
            }
        } catch (e: AEADBadTagException) {
            throw CryptoException.AuthenticationFailed()
        } catch (e: CryptoException) {
            throw e
        } catch (t: Throwable) {
            throw CryptoException.DecryptionFailed("Byte decryption failed", t)
        }
    }

    // ------------------------------------------------------------------
    // Internals
    // ------------------------------------------------------------------

    private fun generateAesKey(): SecretKey {
        val generator = KeyGenerator.getInstance("AES")
        generator.init(CryptoFormat.AES_KEY_LENGTH * 8, secureRandom)
        return generator.generateKey()
    }

    private fun rsaEncryptKey(aesKey: SecretKey, publicKey: PublicKey): ByteArray {
        val cipher = Cipher.getInstance(CryptoFormat.RSA_TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, publicKey)
        return cipher.doFinal(aesKey.encoded)
    }

    private fun rsaDecryptKey(encryptedKey: ByteArray, privateKey: PrivateKey): SecretKey {
        val cipher = Cipher.getInstance(CryptoFormat.RSA_TRANSFORMATION)
        cipher.init(Cipher.DECRYPT_MODE, privateKey)
        val raw = cipher.doFinal(encryptedKey)
        return SecretKeySpec(raw, "AES")
    }

    private fun writeHeader(out: OutputStream, encryptedAesKey: ByteArray, iv: ByteArray) {
        out.write(CryptoFormat.MAGIC_BYTES)
        // NOTE: key-length field is 2 bytes (uint16, big-endian) to stay
        // binary-compatible with the Python reference tool, which writes it
        // with struct.pack(">H", ...). A 2048-bit RSA-OAEP ciphertext is
        // 256 bytes and a 4096-bit one is 512 bytes, both well under the
        // 65535 max of a uint16, so this comfortably covers real key sizes.
        out.write(shortToBigEndianBytes(encryptedAesKey.size))
        out.write(encryptedAesKey)
        out.write(iv)
    }

    /** Reads and validates the header, returning (encryptedAesKey, iv). */
    private fun readHeader(input: InputStream): Pair<ByteArray, ByteArray> {
        val magic = ByteArray(CryptoFormat.MAGIC_BYTES.size)
        readFully(input, magic)
        if (!magic.contentEquals(CryptoFormat.MAGIC_BYTES)) {
            throw CryptoException.InvalidFormat("Unrecognized file header; expected '${CryptoFormat.MAGIC}'")
        }
        val lenBytes = ByteArray(2)
        readFully(input, lenBytes)
        val keyLen = bigEndianBytesToShort(lenBytes)
        if (keyLen <= 0 || keyLen > 4096) {
            throw CryptoException.InvalidFormat("Invalid encrypted key length in header: $keyLen")
        }
        val encryptedKey = ByteArray(keyLen)
        readFully(input, encryptedKey)
        val iv = ByteArray(CryptoFormat.GCM_IV_LENGTH)
        readFully(input, iv)
        return encryptedKey to iv
    }

    /** Streams [input] through [cipher] into [output] using a fixed-size
     * buffer, invoking [onProgress] periodically. Throws [CryptoException.Cancelled]
     * if [token] is cancelled mid-stream. */
    private fun streamThroughCipher(
        input: InputStream,
        output: OutputStream,
        cipher: Cipher,
        totalBytes: Long,
        token: CancellationToken,
        onProgress: ProgressCallback?
    ) {
        val buffer = ByteArray(CryptoFormat.STREAM_BUFFER_SIZE)
        var processed = 0L
        while (true) {
            if (token.isCancelled) throw CryptoException.Cancelled()
            val read = input.read(buffer)
            if (read == -1) break
            val outChunk = cipher.update(buffer, 0, read)
            if (outChunk != null && outChunk.isNotEmpty()) output.write(outChunk)
            processed += read
            onProgress?.invoke(processed, totalBytes)
        }
        if (token.isCancelled) throw CryptoException.Cancelled()
        val finalChunk = cipher.doFinal()
        if (finalChunk.isNotEmpty()) output.write(finalChunk)
        onProgress?.invoke(totalBytes.coerceAtLeast(processed), totalBytes)
    }

    private fun readFully(input: InputStream, buffer: ByteArray) {
        var offset = 0
        while (offset < buffer.size) {
            val read = input.read(buffer, offset, buffer.size - offset)
            if (read == -1) throw CryptoException.InvalidFormat("Unexpected end of stream while reading header")
            offset += read
        }
    }

    /** 2-byte big-endian encoding, matching Python's struct.pack(">H", ...). */
    private fun shortToBigEndianBytes(value: Int): ByteArray = byteArrayOf(
        (value ushr 8 and 0xFF).toByte(),
        (value and 0xFF).toByte()
    )

    private fun bigEndianBytesToShort(bytes: ByteArray): Int =
        ((bytes[0].toInt() and 0xFF) shl 8) or
                (bytes[1].toInt() and 0xFF)

    private fun SecretKey.destroyQuietly() {
        try { if (this is javax.security.auth.Destroyable && !this.isDestroyed) this.destroy() } catch (_: Throwable) {}
    }
}